commitizen init doesn't work with pnpm

I had run the ``commitizen init cz-conventional-changelog --save-dev --save-exact`` command as per README, but it doesn't work, see a copy and paste of the error from commitizen at the end.

Suggestion: when there exists a ``pnpm-lock.yaml`` file, use pnpm instead of npm.

Here's my workaround:

```sh
pnpm add -D -E cz-conventional-changelog
```

Then add this block to the package.json file:

```json
  "config": {
    "commitizen": {
      "path": "cz-conventional-changelog"
    }
  },
```

The ``cz`` command works as usual from this point.

Error from ``commitizen init cz-conventional-changelog --save-dev --save-exact``:


```
Attempting to initialize using the npm package cz-conventional-changelog
npm WARN old lockfile 
npm WARN old lockfile The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile 
npm WARN old lockfile This is a one-time fix-up, please be patient...
npm WARN old lockfile 
npm ERR! code EINVALIDPACKAGENAME
npm ERR! Invalid package name ".pnpm": name cannot start with a period

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/sunnz/.npm/_logs/2021-11-10T12_28_43_823Z-debug.log
Error: Command failed: npm install cz-conventional-changelog --save-dev
npm WARN old lockfile 
npm WARN old lockfile The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile 
npm WARN old lockfile This is a one-time fix-up, please be patient...
npm WARN old lockfile 
npm ERR! code EINVALIDPACKAGENAME
npm ERR! Invalid package name ".pnpm": name cannot start with a period

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/sunnz/.npm/_logs/2021-11-10T12_28_43_823Z-debug.log

    at checkExecSyncError (node:child_process:826:11)
    at Object.execSync (node:child_process:900:15)
    at init (/Users/sunnz/Library/pnpm/nodejs/16.11.1/pnpm-global/5/node_modules/.pnpm/commitizen@4.2.4/node_modules/commitizen/dist/commitizen/init.js:1017:28)
    at Object.bootstrap (/Users/sunnz/Library/pnpm/nodejs/16.11.1/pnpm-global/5/node_modules/.pnpm/commitizen@4.2.4/node_modules/commitizen/dist/cli/commitizen.js:34:30)
    at Object.<anonymous> (/Users/sunnz/Library/pnpm/nodejs/16.11.1/pnpm-global/5/node_modules/.pnpm/commitizen@4.2.4/node_modules/commitizen/bin/commitizen.js:2:38)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19) {
  status: 1,
  signal: null,
  output: [
    null,
    <Buffer >,
    <Buffer 6e 70 6d 20 57 41 52 4e 20 6f 6c 64 20 6c 6f 63 6b 66 69 6c 65 20 0a 6e 70 6d 20 57 41 52 4e 20 6f 6c 64 20 6c 6f 63 6b 66 69 6c 65 20 54 68 65 20 70 ... 491 more bytes>
  ],
  pid: 55718,
  stdout: <Buffer >,
  stderr: <Buffer 6e 70 6d 20 57 41 52 4e 20 6f 6c 64 20 6c 6f 63 6b 66 69 6c 65 20 0a 6e 70 6d 20 57 41 52 4e 20 6f 6c 64 20 6c 6f 63 6b 66 69 6c 65 20 54 68 65 20 70 ... 491 more bytes>
}
```

--- a/README.md
+++ b/README.md
@@ -58,13 +58,14 @@ npm install commitizen -g
 Next, initialize your project to use the cz-conventional-changelog adapter by typing:
 
 ```sh
+# npm
 commitizen init cz-conventional-changelog --save-dev --save-exact
-```
 
-Or if you are using Yarn:
-
-```sh
+# yarn
 commitizen init cz-conventional-changelog --yarn --dev --exact
+
+# pnpm
+commitizen init cz-conventional-changelog --pnpm --save-dev --save-exact
 ```
 
 Note that if you want to force install over the top of an old adapter, you can apply the `--force` argument. For more information on this, just run `commitizen help`.

--- a/src/commitizen/adapter.js
+++ b/src/commitizen/adapter.js
@@ -11,12 +11,10 @@ export {
   addPathToAdapterConfig,
   getNearestNodeModulesDirectory,
   getNearestProjectRootDirectory,
-  getNpmInstallStringMappings,
+  getInstallStringMappings,
   getPrompter,
-  generateNpmInstallAdapterCommand,
+  generateInstallAdapterCommand,
   resolveAdapterPath,
-  getYarnAddStringMappings,
-  generateYarnAddAdapterCommand,
   getGitRootPath,
 };
 
@@ -55,40 +53,32 @@ function addPathToAdapterConfig (cliPath, repoPath, adapterNpmName) {
   fs.writeFileSync(packageJsonPath, JSON.stringify(newPackageJsonContent, null, indent) + '\n');
 }
 
-/**
- * Generates an npm install command given a map of strings and a package name
+/*
+ * Get additional options for install command
  */
-function generateNpmInstallAdapterCommand (stringMappings, adapterNpmName) {
+function getInstallOptions(stringMappings) {
+  return Array.from(stringMappings.values()).filter(Boolean).join(" ")
+}
 
-  // Start with an initial npm install command
-  let installAdapterCommand = `npm install ${adapterNpmName}`;
+/*
+ * Get specific install command for passed package manager
+ */
+function getInstallCommand(packageManager) {
+  const fallbackCommand = 'install';
+  const commandByPackageManager = {
+    npm: 'install',
+    yarn: 'add',
+    pnpm: 'add',
+  };
 
-  // Append the neccesary arguments to it based on user preferences
-  for (let value of stringMappings.values()) {
-    if (value) {
-      installAdapterCommand = installAdapterCommand + ' ' + value;
-    }
-  }
-
-  return installAdapterCommand;
+  return commandByPackageManager[packageManager] || fallbackCommand;
 }
 
 /**
- * Generates an yarn add command given a map of strings and a package name
+ * Generates an npm install command given a map of strings and a package name
  */
-function generateYarnAddAdapterCommand (stringMappings, adapterNpmName) {
-
-  // Start with an initial yarn add command
-  let installAdapterCommand = `yarn add ${adapterNpmName}`;
-
-  // Append the necessary arguments to it based on user preferences
-  for (let value of stringMappings.values()) {
-    if (value) {
-      installAdapterCommand = installAdapterCommand + ' ' + value;
-    }
-  }
-
-  return installAdapterCommand;
+function generateInstallAdapterCommand(stringMappings, adapterNpmName, packageManager = "npm") {
+  return `${packageManager} ${getInstallCommand(packageManager)} ${adapterNpmName} ${getInstallOptions(stringMappings)}`;
 }
 
 /**
@@ -117,24 +107,28 @@ function getNearestProjectRootDirectory (repoPath, options) {
 }
 
 /**
- * Gets a map of arguments where the value is the corresponding npm strings
+ * Gets a map of arguments where the value is the corresponding (to passed package manager) string
  */
-function getNpmInstallStringMappings (save, saveDev, saveExact, force) {
-  return new Map()
-    .set('save', (save && !saveDev) ? '--save' : undefined)
+function getInstallStringMappings({ save, dev, saveDev, exact, saveExact, force }, packageManager) {
+  const npm = new Map()
+    .set('save', save && !saveDev ? '--save' : undefined)
     .set('saveDev', saveDev ? '--save-dev' : undefined)
     .set('saveExact', saveExact ? '--save-exact' : undefined)
     .set('force', force ? '--force' : undefined);
-}
 
-/**
- * Gets a map of arguments where the value is the corresponding yarn strings
- */
-function getYarnAddStringMappings (dev, exact, force) {
-  return new Map()
+  const yarn = new Map()
     .set('dev', dev ? '--dev' : undefined)
     .set('exact', exact ? '--exact' : undefined)
     .set('force', force ? '--force' : undefined);
+
+  const pnpm = new Map()
+    .set('save', save && !saveDev ? '--save-prod' : undefined)
+    .set('dev', saveDev ? '--save-dev' : undefined)
+    .set('exact', saveExact ? '--save-exact' : undefined);
+
+  const map = { npm, yarn, pnpm };
+
+  return map[packageManager] || npm;
 }
 
 /**

--- a/src/commitizen/init.js
+++ b/src/commitizen/init.js
@@ -5,10 +5,8 @@ import * as adapter from './adapter';
 
 let {
   addPathToAdapterConfig,
-  generateNpmInstallAdapterCommand,
-  getNpmInstallStringMappings,
-  generateYarnAddAdapterCommand,
-  getYarnAddStringMappings,
+  generateInstallAdapterCommand,
+  getInstallStringMappings,
 } = adapter;
 
 export default init;
@@ -43,6 +41,8 @@ const defaultInitOptions = {
   yarn: false,
   dev: true,
   exact: false, // should add trailing comma, thus next developer doesn't got blamed for this line
+
+  pnpm: false, // reuses `save`, `saveDev`, `saveExact`
 };
 
 /**
@@ -56,6 +56,7 @@ function init (repoPath, adapterNpmName, {
   yarn = false,
   dev = false,
   exact = false,
+  pnpm = false,
   includeCommitizen = false
 } = defaultInitOptions) {
 
@@ -65,13 +66,15 @@ function init (repoPath, adapterNpmName, {
   // Load the current adapter config
   let adapterConfig = loadAdapterConfig(repoPath);
 
+  const packageManager = yarn ? 'yarn' : pnpm ? 'pnpm' : 'npm';
+
   // Get the npm string mappings based on the arguments provided
-  let stringMappings = yarn ? getYarnAddStringMappings(dev, exact, force) : getNpmInstallStringMappings(save, saveDev, saveExact, force);
+  const stringMappings = getInstallStringMappings({ save, dev, saveDev, saveExact, force }, packageManager);
 
   // Generate a string that represents the npm install command
-  let installAdapterCommand = yarn ? generateYarnAddAdapterCommand(stringMappings, adapterNpmName) : generateNpmInstallAdapterCommand(stringMappings, adapterNpmName);
+  const installAdapterCommand = generateInstallAdapterCommand(stringMappings, adapterNpmName, packageManager);
 
-  let installCommitizenCommand = yarn ? generateYarnAddAdapterCommand(stringMappings, "commitizen") : generateNpmInstallAdapterCommand(stringMappings, "commitizen");
+  const installCommitizenCommand = generateInstallAdapterCommand(stringMappings, 'commitizen', packageManager);
 
   // Check for previously installed adapters
   if (adapterConfig && adapterConfig.path && adapterConfig.path.length > 0 && !force) {

--- a/test/tests/init.js
+++ b/test/tests/init.js
@@ -173,130 +173,140 @@ describe('init', function () {
 
   });
 
-  it('installs an adapter with --yarn', function () {
+  const supportedPackageManagers = ['yarn', 'pnpm'];
 
-    this.timeout(config.maxTimeout); // this could take a while
+  supportedPackageManagers.forEach((packageManger) => {
+    describe(`alternative package managers: ${packageManger}`, () => {
+      it(`installs an adapter with ${packageManger}`, function () {
+        this.timeout(config.maxTimeout); // this could take a while
 
-    // SETUP
+        // SETUP
 
-    // Install an adapter
-    commitizenInit(config.paths.endUserRepo, 'cz-conventional-changelog', { yarn: true });
+        // Install an adapter
+        commitizenInit(config.paths.endUserRepo, 'cz-conventional-changelog', { [packageManger]: true });
 
-    // TEST
+        // TEST
 
-    // Check resulting json
-    let packageJson = util.getParsedPackageJsonFromPath(config.paths.endUserRepo);
-    expect(packageJson).to.have.nested.property('dependencies.cz-conventional-changelog');
+        // Check resulting json
+        let packageJson = util.getParsedPackageJsonFromPath(config.paths.endUserRepo);
+        if (packageManger === 'yarn') {
+          expect(packageJson).to.have.nested.property('dependencies.cz-conventional-changelog');
+        } else {
+          expect(packageJson).to.have.nested.property('devDependencies.cz-conventional-changelog');
+        }
+      });
 
-  });
+      it(`installs an adapter with ${packageManger} --dev`, function () {
+        this.timeout(config.maxTimeout); // this could take a while
 
-  it('installs an adapter with --yarn --dev', function () {
+        // SETUP
 
-    this.timeout(config.maxTimeout); // this could take a while
+        // Install an adapter
+        commitizenInit(config.paths.endUserRepo, 'cz-conventional-changelog', { [packageManger]: true, dev: true });
 
-    // SETUP
+        // TEST
 
-    // Install an adapter
-    commitizenInit(config.paths.endUserRepo, 'cz-conventional-changelog', { yarn: true, dev: true });
+        // Check resulting json
+        let packageJson = util.getParsedPackageJsonFromPath(config.paths.endUserRepo);
+        expect(packageJson).to.have.nested.property('devDependencies.cz-conventional-changelog');
+      });
 
-    // TEST
+      it(`errors (with ${packageManger}) on previously installed adapter`, function () {
+        this.timeout(config.maxTimeout); // this could take a while
 
-    // Check resulting json
-    let packageJson = util.getParsedPackageJsonFromPath(config.paths.endUserRepo);
-    expect(packageJson).to.have.nested.property('devDependencies.cz-conventional-changelog');
+        // SETUP
 
-  });
+        // Add a first adapter
+        commitizenInit(config.paths.endUserRepo, 'cz-conventional-changelog', { [packageManger]: true, dev: true });
 
-  it('errors (with --yarn) on previously installed adapter', function () {
+        // TEST
+        // Adding a second adapter
+        expect(function () {
+          commitizenInit(config.paths.endUserRepo, 'cz-jira-smart-commit', { [packageManger]: true, dev: true });
+        }).to.throw(/already configured/);
 
-    this.timeout(config.maxTimeout); // this could take a while
+        // Check resulting json
+        let packageJson = util.getParsedPackageJsonFromPath(config.paths.endUserRepo);
+        expect(packageJson).not.to.have.nested.property('devDependencies', 'cz-jira-smart-commit');
+        expect(packageJson).to.have.nested.property(
+          'config.commitizen.path',
+          './node_modules/cz-conventional-changelog'
+        );
+        // TODO: Eventually may need to offer both path and package keys. package = npm package name
+        // Path for local development
+      });
 
-    // SETUP
+      it(`succeeds (with ${packageManger}) if force is true`, function () {
+        this.timeout(config.maxTimeout); // this could take a while
 
-    // Add a first adapter
-    commitizenInit(config.paths.endUserRepo, 'cz-conventional-changelog', { yarn: true, dev: true });
+        // SETUP
 
-    // TEST
-    // Adding a second adapter
-    expect(function () {
-      commitizenInit(config.paths.endUserRepo, 'cz-jira-smart-commit', { yarn: true, dev: true });
-    }).to.throw(/already configured/);
+        // Add a first adapter
+        commitizenInit(config.paths.endUserRepo, 'cz-conventional-changelog', { [packageManger]: true, dev: true });
 
-    // Check resulting json
-    let packageJson = util.getParsedPackageJsonFromPath(config.paths.endUserRepo);
-    expect(packageJson).not.to.have.nested.property('devDependencies', 'cz-jira-smart-commit');
-    expect(packageJson).to.have.nested.property('config.commitizen.path', './node_modules/cz-conventional-changelog');
-    // TODO: Eventually may need to offer both path and package keys. package = npm package name
-    // Path for local development
-  });
+        // TEST
 
-  it('succeeds (with --yarn) if force is true', function () {
+        // Adding a second adapter
+        expect(function () {
+          commitizenInit(config.paths.endUserRepo, 'cz-jira-smart-commit', {
+            [packageManger]: true,
+            dev: true,
+            force: true,
+          });
+        }).to.not.throw();
 
-    this.timeout(config.maxTimeout); // this could take a while
+        let packageJson = util.getParsedPackageJsonFromPath(config.paths.endUserRepo);
+        expect(packageJson.devDependencies).to.have.property('cz-jira-smart-commit');
+        expect(packageJson).to.have.nested.property('config.commitizen.path', './node_modules/cz-jira-smart-commit');
+      });
 
-    // SETUP
+      it(`installs (with ${packageManger}) an adapter without --save-exact`, function () {
+        this.timeout(config.maxTimeout); // this could take a while
 
-    // Add a first adapter
-    commitizenInit(config.paths.endUserRepo, 'cz-conventional-changelog', { yarn: true, dev: true });
+        // SETUP
 
-    // TEST
+        // Add a first adapter
+        commitizenInit(config.paths.endUserRepo, 'cz-conventional-changelog', { [packageManger]: true, dev: true });
+        let packageJson = util.getParsedPackageJsonFromPath(config.paths.endUserRepo);
 
-    // Adding a second adapter
-    expect(function () {
-      commitizenInit(config.paths.endUserRepo, 'cz-jira-smart-commit', { yarn: true, dev: true, force: true });
-    }).to.not.throw();
+        // TEST
+        expect(packageJson.devDependencies).to.have.property('cz-conventional-changelog');
+        let range = packageJson.devDependencies['cz-conventional-changelog'];
 
-    let packageJson = util.getParsedPackageJsonFromPath(config.paths.endUserRepo);
-    expect(packageJson.devDependencies).to.have.property('cz-jira-smart-commit');
-    expect(packageJson).to.have.nested.property('config.commitizen.path', './node_modules/cz-jira-smart-commit');
+        // It should satisfy the requirements of a range
+        expect(semver.validRange(range)).to.not.equal(null);
 
-  });
+        // // But you CAN NOT increment a range
+        // expect(semver.inc(range, 'major')).to.equal(null);
+        // TODO: We need to figure out how to check if the repo has save exact set
+        // in the config before we can re-enable this. The --save-exact setting
+        // in our package.json breaks this test
+      });
 
-  it('installs (with --yarn) an adapter without --save-exact', function () {
+      it(`installs an adapter with ${packageManger} --exact`, function () {
+        this.timeout(config.maxTimeout); // this could take a while
 
-    this.timeout(config.maxTimeout); // this could take a while
+        // SETUP
 
-    // SETUP
+        // Add a first adapter
+        commitizenInit(config.paths.endUserRepo, 'cz-conventional-changelog', {
+          [packageManger]: true,
+          dev: true,
+          exact: true,
+        });
+        let packageJson = util.getParsedPackageJsonFromPath(config.paths.endUserRepo);
 
-    // Add a first adapter
-    commitizenInit(config.paths.endUserRepo, 'cz-conventional-changelog', { yarn: true, dev: true });
-    let packageJson = util.getParsedPackageJsonFromPath(config.paths.endUserRepo);
+        // TEST
+        expect(packageJson.devDependencies).to.have.property('cz-conventional-changelog');
+        let range = packageJson.devDependencies['cz-conventional-changelog'];
 
-    // TEST
-    expect(packageJson.devDependencies).to.have.property('cz-conventional-changelog');
-    let range = packageJson.devDependencies['cz-conventional-changelog'];
+        // It should satisfy the requirements of a range
+        expect(semver.validRange(range)).to.not.equal(null);
 
-    // It should satisfy the requirements of a range
-    expect(semver.validRange(range)).to.not.equal(null);
-
-    // // But you CAN NOT increment a range
-    // expect(semver.inc(range, 'major')).to.equal(null);
-    // TODO: We need to figure out how to check if the repo has save exact set
-    // in the config before we can re-enable this. The --save-exact setting
-    // in our package.json breaks this test
-
-  });
-
-  it('installs an adapter with --yarn --exact', function () {
-
-    this.timeout(config.maxTimeout); // this could take a while
-
-    // SETUP
-
-    // Add a first adapter
-    commitizenInit(config.paths.endUserRepo, 'cz-conventional-changelog', { yarn: true, dev: true, exact: true });
-    let packageJson = util.getParsedPackageJsonFromPath(config.paths.endUserRepo);
-
-    // TEST
-    expect(packageJson.devDependencies).to.have.property('cz-conventional-changelog');
-    let range = packageJson.devDependencies['cz-conventional-changelog'];
-
-    // It should satisfy the requirements of a range
-    expect(semver.validRange(range)).to.not.equal(null);
-
-    // But you CAN increment a single version
-    expect(semver.inc(range, 'major')).not.to.equal(null);
-
+        // But you CAN increment a single version
+        expect(semver.inc(range, 'major')).not.to.equal(null);
+      });
+    });
   });
 
 });

